design-doc.md

```diff
@@ -0,0 +1,128 @@
+# Sherlock Design Doc
+
+## Terminology
+
+A **stream** is a persistent, sequential, structured, immutable, append-only storage of data.
+
+The quantum storage unit of the stream is an **entry**.
+
+Entries can be appended to streams using **publishers** and scanned with **listeners**.
+
+
+## Entries
+
+Streams are strongly typed, and within a stream, each entry has the type specified at stream declaration time. All entries within one stream have the same type. The type can, and often is, polymorphic.
+
+Implementation-wise, entries of polymorphic type are `std::unique_ptr<>`-s of `SomeBaseType`: the user they call their virtual functions or have individual instances of `[const] SomeBaseType&` dispatched via other RTTI mechanisms.
+
+The type of entries in the stream is defined at stream declaration time. When serializing/deserealizing streams to/from disk, a basic check of the signature of storage schema is performed to verify that no data corruption may occur.
+
+Polymorphic types can be extended with more derived types, but it is the user's responsibilty that no binary attempts to deserialize entries of the type not supported by this particular version. (`TODO(dkorolev): See how strict we can be with this in V1.`)
+
+In order to qualify for a stream type, it should be guaranteed that an entry of this type can be serialized and de-serialized, and that the very same entry will be constructed as a result of the serialization/de-serialization cycle. In other words, stream entries should be agnostic with respect to serialization. This is a requirement for dropping strict guarantees on the number of serializations that take place internally.
```

Comment:
I'd suggest to remove a hyphen from `de-serialized` & `de-serialization` for consistency